src/FrequencyBalancerAgent.cpp

```diff
@@ -0,0 +1,644 @@
+/*
+ * Copyright (c) 2015 - 2023, Intel Corporation
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include "config.h"
+
+#include "FrequencyBalancerAgent.hpp"
+
+#include <unistd.h>
+
+#include "Environment.hpp"
+#include "FrequencyGovernor.hpp"
+#include "FrequencyLimitDetector.hpp"
+#include "FrequencyTimeBalancer.hpp"
+#include "PowerGovernor.hpp"
+#include "SSTClosGovernor.hpp"
+#include "Waiter.hpp"
+#include "geopm/Exception.hpp"
+#include "geopm/Helper.hpp"
+#include "geopm/PlatformIO.hpp"
+#include "geopm/PlatformTopo.hpp"
+#include "geopm_debug.hpp"
+#include "geopm_hash.h"
+#include "geopm_hint.h"
+#include "geopm_time.h"
+#include "geopm_topo.h"
+
+#include <algorithm>
+#include <chrono>
+#include <cmath>
+#include <exception>
+#include <iterator>
+#include <numeric>
+#include <thread>
+#include <utility>
+
+// Minimum number of sampling wait periods before applying new epoch controls.
+#define MINIMUM_WAIT_PERIODS_FOR_NEW_EPOCH_CONTROL 5
```

Comment:
Why #define's instead of constexpr's?  Please use constexpr's in the header file unless these are needed for some other reason.